libtock/alarm_timer.c

```diff
@@ -106,6 +131,84 @@ int alarm_at(uint32_t reference, uint32_t dt, subscribe_upcall cb, void* ud, ala
   return RETURNCODE_SUCCESS;
 }
 
+// Holds original callback and user data. Also keeps bookkeeping for the
+// intermediate alarm overflows.
+typedef struct overflow_ud {
+  // How many times the alarm has left to overflow before reaching target time
+  int overflows_left;
+  // Number of remaining ticks after last full alarm overflow
+  int remainder_ticks;
+  // Original user data passed to `timer_in`
+  void* original_ud;
+  // Original callback passed to `timer_in`
+  subscribe_upcall* original_cb;
+  alarm_t* alarm;
+} overflow_ud_t;
+
+// The intermediate callback that handles overflows of alarm. This is used by
+// `timer_in` to handle cases where `alarm_at` would normally be unable to
+// support the whole timer length. `alarm_at` can only keep track of up to 2^32
+// ticks.
+static void overflow_callback(int                          last_timer_fire_time,
+                              __attribute__ ((unused)) int unused1,
+                              __attribute__ ((unused)) int unused2,
+                              void*                        overflow_ud) {
+  overflow_ud_t* ud = (overflow_ud_t*)overflow_ud;
+
+  if (ud->overflows_left == 0) {
+    // no overflows left, schedule last alarm with original callback
+    alarm_at(
+      last_timer_fire_time,
+      ud->remainder_ticks,
+      ud->original_cb,
+      ud->original_ud,
+      ud->alarm
+      );
+  } else {
+    // schedule next intermediate alarm that will overflow
+    ud->overflows_left--;
+
+    alarm_at(
+      last_timer_fire_time,
+      MAX_TICKS,
+      (subscribe_upcall*) overflow_callback,
+      (void*) ud,
+      ud->alarm
+      );
+  }
+}
+
+// Timer implementation
+int timer_in(uint32_t dt_ms, subscribe_upcall cb, void* ud, tock_timer_t *tock_timer) {
+  alarm_t* alarm = &(tock_timer->alarm);
+  uint32_t now;
+  int ret = alarm_internal_read(&now);
+  if (ret != RETURNCODE_SUCCESS) return ret;
+
+  // If `dt_ms` is longer than the time that an alarm can count up to, then `timer_in` will
+  // schedule multiple alarms to reach the full length. We calculate the number of full overflows
+  // and the remainder ticks to reach the target length of time. The overflows use the
+  // `overflow_callback` for each intermediate overflow.
+  uint32_t max_ticks_in_ms = ticks_to_ms(MAX_TICKS);
+  if (dt_ms > max_ticks_in_ms) {
```

Comment:
We should report this back to the user somehow. This could have a large impact on power usage for long sleeping devices as we wake up the CPU a lot more then we should. Maybe a `bool` to enable this feature?

Reply:
For the common case of alarms less than 2^32 ticks (which is about 36 hrs for a 32KHz clock), no additional alarms will be scheduled. The behavior will match how `timer_in` worked before these changes. We've verified this with our testing program (found at `examples/test/timer_overflow/main.c`) and it can be reproduced by setting `TIMER_LEN_MS` to a length of time that will not overflow the counter.

For the uncommon case of alarms greater than 2^32 ticks there will be additional overhead. In our testing we have observed it to be around 1ms. This can be reproduced by setting `MAX_TICKS` in `libtock/alarm_timer.c` to something smaller than `TIMER_LEN_MS` in our testing file. For our Nordic NRF52840 board with an oscillator of 32768Hz, we set  `MAX_TICKS` to 32768Hz which will simulate overflowing every 1 second. In our testing file we set an alarm of 2 seconds (setting `TIMER_LEN_MS` to 2000) which means that we will schedule two 1 second long alarms to reach our target time of 2 seconds. As mentioned, the overhead of scheduling an additional intermediate alarm is only 1ms. This is tolerable since this will only happen when setting long running timers (over 36 hours with the 32KHz oscillator).

Reply:
Also, with regards to waking up the CPU more often. The CPU will not be waken up any more than the previous version of `timer_in` for the common case of alarms less than 2^32 ticks. 

The alternative to having this feature is that `timer_in` will behave incorrectly when the user tries to set a long running alarm that will overflow the counter. We think that by default `timer_in` should count the correct amount of time.